### ISSUE TITLE
provisioner/chef: unregister the client from chef server when replacing an instance

### ISSUE THREAD

**Issue by pmvilaca** [hashicorp/terraform#3605]
It would be useful to get terraform deleting the client from chef server for the instances that are being replaced. Otherwise, the replacement operation will always fail during the provisioning task because the node already exists on the chef server (assuming that you're defining the node_name)

Example:

```
provisioner "chef"  {
    run_list = ["role[common]"]
    node_name = "nat_instance_ffm_1a"
   ...
```

Result:
```
aws_instance.nat_1b (chef): ================================================================================
aws_instance.nat_1b (chef): Chef encountered an error attempting to create the client "nat_instance_ffm_1b"
aws_instance.nat_1b (chef): ================================================================================
```

**Comment by partamonov** [comment]
I guess this is related to #649 #2010 #2720. And I agree this is nice to have.

**Comment by apparentlymart** [comment]
I left [a comment about this](https://github.com/hashicorp/terraform/issues/386#issuecomment-154596719) in #386 way back in the days of three-digit issues. :grinning:

Right now this sort of thing is not supported by Terraform's architecture since it doesn't allow for actions on delete, but that issue #386 is about supporting that, which could then in turn solve this more-specific problem of leaving junk nodes in Chef Server. In our case we were naming the nodes after the EC2 instance id and so the outcome was hundreds of defunct nodes rather than a naming conflict, but still annoying especially since paid chef uses "number of nodes" as a quota metric.

**Comment by bitglue** [comment]
> not supported by Terraform's architecture since it doesn't allow for actions on delete

Isn't the destroy action on a resource exactly that?

It would seem to me that what's missing is a Terraform resource that represents the membership in a chef server. Then you not only have a way to add clients, and remove clients, but also reconcile state and recover from failures.

**Comment by rberger** [comment]
Does it look like this is something anyone is working on? Its a huge pain point at least for me and I would assume anyone who is using Terraform and Chef.

**Comment by svanharmelen** [comment]
@rberger no I don't think so unfortunately... It's a thing that needs to be added to Terraform core and I didn't had time to have a look at it lately...

**Comment by orclev** [comment]
Just voicing my wish for this as well, currently before we can run a terraform apply we need to manually go in and delete all the registered clients in chef.

**Comment by svanharmelen** [comment]
This should be fixed by PR #8577 which introduces a new attribute called `recreate_client`. If set to `true`, it will first delete the existing node and client before registering the new one.
